Return nonzero exit code if there are errors at a stop point

fhahn opened the pull request:
At the moment, when compilation is stopped at a stop point (like `-Z parse-only`), `rustc` does not return an nonzero exit code even if there are errors (expect fatal ones, that cause it to panic immediately). As an example, compiling `src/test/compile-fail/doc-before-semi.rs` with `-Z parse-only` raises an error, but exists with 0.

Note that I could not use `sess.abort_if_errors()` in the macro, because `sess` is passed by value and move at some point.

rust-highfive commented:
r? @nikomatsakis

(rust_highfive has picked a reviewer for you, use r? to override)

nikomatsakis commented:
Can you add some kind of test for this?

brson commented:
@nikomatsakis https://github.com/rust-lang/rust/pull/22118 contains a bunch of test cases for this. Since those are the intended use it may be hard to dig up other relevant test cases.

fhahn commented:
@nikomatsakis @brson yes, I would like to add the `-Z parse-only` flag to all `parse-fail` tests after #22118 lands to test this patch. Also I am going to check if there are test cases that fail at other stop points with this patch.

fhahn commented:
@brson I have updated the patch to use a straight forward solution to invoke the diagnostics handler. I added a session argument to the `controller_entry_point` macro to pass the current valid session. This way we can handle the case were `session` [got moved](https://github.com/rust-lang/rust/pull/22117/files#diff-698522bf32a92da4331b4bd3f6ed31c6R128).

fhahn commented:
@brson what do you think about the new approach? Is it more reasonable?

pczarn commented:
Needs rebase

fhahn commented:
Thanks, I have rebased the PR.

bors commented:
:umbrella: The latest upstream changes (presumably #23331) made this pull request unmergeable. Please resolve the merge conflicts.

fhahn commented:
I have updated this pull request.

bors commented:
:umbrella: The latest upstream changes (presumably #23654) made this pull request unmergeable. Please resolve the merge conflicts.

fhahn commented:
rebased this PR again

bors commented:
:umbrella: The latest upstream changes (presumably #23930) made this pull request unmergeable. Please resolve the merge conflicts.

tamird commented on src/test/parse-fail/attr-before-eof.rs:
```diff
@@ -8,4 +8,6 @@
 // option. This file may not be copied, modified, or distributed
 // except according to those terms.
 
+// compile-flags: -Z parse-only
```
can this be globally added to `parse-fail` rather than in each test file?

nikomatsakis commented:
r+

@fhahn I just realized you've been dutifully rebasing this PR and I didn't notice at all. I apologize for that. In any case, I think it seems clearly good!

nikomatsakis commented:
(Unfortunately, it does need to be rebased again...)

fhahn commented:
@nikomatsakis no worries, all rebases were very minimal & straightforward

pnkfelix commented:
@bors r=nikomatsakis

bors commented:
:pushpin: Commit 2a24e97 has been approved by `nikomatsakis`

<!-- @bors r=nikomatsakis 2a24e97c80d5004cd1b57cb96a0bde6fe43a848d -->

bors commented:
:hourglass: Testing commit 2a24e97 with merge be9a72b...

bors commented:
:sunny: Test successful - [auto-linux-32-nopt-t](http://buildbot.rust-lang.org/builders/auto-linux-32-nopt-t/builds/4590), [auto-linux-32-opt](http://buildbot.rust-lang.org/builders/auto-linux-32-opt/builds/4597), [auto-linux-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-linux-64-nopt-t/builds/4561), [auto-linux-64-opt](http://buildbot.rust-lang.org/builders/auto-linux-64-opt/builds/4578), [auto-linux-64-x-android-t](http://buildbot.rust-lang.org/builders/auto-linux-64-x-android-t/builds/4546), [auto-mac-32-opt](http://buildbot.rust-lang.org/builders/auto-mac-32-opt/builds/4612), [auto-mac-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-mac-64-nopt-t/builds/4610), [auto-mac-64-opt](http://buildbot.rust-lang.org/builders/auto-mac-64-opt/builds/4589), [auto-win-32-nopt-t](http://buildbot.rust-lang.org/builders/auto-win-32-nopt-t/builds/4278), [auto-win-32-opt](http://buildbot.rust-lang.org/builders/auto-win-32-opt/builds/4265), [auto-win-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-win-64-nopt-t/builds/3083), [auto-win-64-opt](http://buildbot.rust-lang.org/builders/auto-win-64-opt/builds/3102)
